File: 713. Subarray Product Less Than K/optimised.py
#!/usr/bin/env python

from typing import List

def numSubarrayProductLessThanK(nums: List[int], k: int) -> int:
    if k < 1:
        return 0

    product = 1
    count = 0
    i = 0
    for j in range(len(nums)):
        product *= nums[j]

        while i <= j and product >= k:
            product //= nums[i]
            i += 1

        count += j - i + 1

    return count

print(numSubarrayProductLessThanK([10,5,2,6], 100))
print(numSubarrayProductLessThanK([1,2,3], 0))

File: 713. Subarray Product Less Than K/optimised.js
"use strict";
function numSubarrayProductLessThanK(nums, k) {
    const n = nums.length;
    let count = 0;
    let product = 1;
    let i = 0;
    for (let j = 0; j < n; j++) {
        product *= nums[j];
        while (product >= k) {
            product = Math.floor(product / nums[j]);
        }
        count += j - i + 1;
    }
    return count;
}
;
console.log(numSubarrayProductLessThanK([10, 5, 2, 6], 100));
console.log(numSubarrayProductLessThanK([1, 2, 3], 0));
